Add migration steps for JWT/OAuth/OpenID Connect/Certificates (#403)
SteeltoeOSS/Documentation: docs/docs/v4/welcome/migrate-quick-steps.md

The CredHub client has been removed from Steeltoe in v4.
Use [CredHub Service Broker](https://techdocs.broadcom.com/us/en/vmware-tanzu/platform-services/credhub-service-broker/services/credhub-sb/index.html) instead.

### OAuth / OpenID Connect

Project file:

```diff
<Project>
  <ItemGroup>
-    <PackageReference Include="Steeltoe.Extensions.Configuration.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Configuration.CloudFoundry" Version="4.0.0" />
-    <PackageReference Include="Steeltoe.Security.Authentication.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Security.Authentication.OpenIdConnect" Version="4.0.0" />
  </ItemGroup>
</Project>
```

> Program.cs:

Reviewer: Running the 3x code results in an exception.

```
System.InvalidOperationException
  HResult=0x80131509
  Message=Scheme already exists: CloudFoundry
  Source=Microsoft.AspNetCore.Authentication.Abstractions
  StackTrace:
   at Microsoft.AspNetCore.Authentication.AuthenticationOptions.AddScheme(String name, Action`1 configureBuilder)
   at Microsoft.Extensions.Options.OptionsFactory`1.Create(String name)
   at Microsoft.Extensions.Options.UnnamedOptionsManager`1.get_Value()
   at Microsoft.AspNetCore.Authentication.AuthenticationSchemeProvider..ctor(IOptions`1 options, IDictionary`2 schemes)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
   at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(ServiceIdentifier serviceIdentifier)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.Internal.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider)
   at Microsoft.Extensions.Internal.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.ReflectionMiddlewareBinder.CreateMiddleware(RequestDelegate next)
   at Microsoft.AspNetCore.Builder.ApplicationBuilder.Build()
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.<StartAsync>d__40.MoveNext()
   at Microsoft.Extensions.Hosting.Internal.Host.<<StartAsync>b__15_1>d.MoveNext()
   at Microsoft.Extensions.Hosting.Internal.Host.<ForeachService>d__18`1.MoveNext()
   at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>d__15.MoveNext()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.<RunAsync>d__4.MoveNext()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.<RunAsync>d__4.MoveNext()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at Program.<Main>$(String[] args) in C:\Source\Projects\SteeltoeMigrate\Program.cs:line 136
```

Reviewer: This still doesn't work in 3x. To test, I've added the following after `app.UseAuthorization()`:

```c#
app.MapGet("/ping", async httpContext =>
{
    httpContext.Response.StatusCode = 200;
    httpContext.Response.ContentType = "text/plain";
    await httpContext.Response.WriteAsync("pong");
}).RequireAuthorization();
```

Sending an HTTP or HTTPS request returns:

> System.InvalidOperationException: IDX20803: Unable to obtain configuration from: '[PII is hidden. For more details, see https://aka.ms/IdentityModel/PII.]'.
> ...
> System.Net.Http.HttpRequestException: No such host is known. (default_oauthserviceurl:443)

After replacing `.AddCloudFoundryOpenIdConnect(builder.Configuration);` with `.AddCloudFoundryOAuth(builder.Configuration);`, it returns:

> No such host is known. (default_oauthserviceurl:80)

Author: Yeah, 3.x never worked with local UAA... I'll open a PR to add support for config-mapping of these additional properties, but as-is this would be required:

```
authbuilder.AddCloudFoundryOpenIdConnect(CloudFoundryDefaults.AuthenticationScheme, builder.Configuration, (options, _) =>
{
    options.MetadataAddress = "http://localhost:8080/.well-known/openid-configuration";
    options.RequireHttpsMetadata = false;
});
```

Reviewer: Running against https://github.com/SteeltoeOSS/Steeltoe/pull/1574, this now _always_ grants access, despite putting an invalid `client_secret` in configuration. I've tried adding `.RequireAuthorization("some-group")` to the `ping` endpoint, but that doesn't resolve the issue. Am I doing something wrong?

Author: It could be hard to say without looking at the whole app (or at least the definition of the authorization policy). If it still isn't working after the latest updates, refer to the [apps I used through this process](https://github.com/SteeltoeOSS/Steeltoe/pull/1574#issuecomment-3233794873) or provide the broken app

Reviewer: Fair enough. I'll give it another try, based on all the updates elsewhere.

Reviewer: #### TEST OUTCOME
(based on `3.3.0-3-x-00226` and `4.0.0-main-00225`)

- :white_check_mark: OAuth: 3x http/https with/without RequireAuthorization
- :white_check_mark: OIDC: 3x/4x http/https with/without RequireAuthorization
- :white_check_mark: JWT: 3x/4x http/https with/without RequireAuthorization
- :white_check_mark: mTLS server-to-server: 3x/4x http/https with/without RequireAuthorization
- :white_check_mark: mTLS client-to-server: 3x-to-3x/3x-to-4x/4x-to-3x/4x-to-4x http/https with/without RequireAuthorization (regen all certs each time, sleep 2s in client to avoid gen clash)

Except for mTLS, I haven't verified that access is _granted_ with the proper token. Please let me know if I should test additional cases. It remains an interesting case to verify that access is denied with an invalid secret in configuration, but I'm not sure how to set that up (I've used `.http` files, so far).

Author: I don't think there are any additional cases that need further testing at this point.

> ```diff

Reviewer: The code compiles for 4x, but any request (http or https, actuator or weatherforecast) always fails with the following error:

```
System.InvalidOperationException: The MetadataAddress or Authority must use HTTPS unless disabled for development by setting RequireHttpsMetadata=false.
   at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectPostConfigureOptions.PostConfigure(String name, OpenIdConnectOptions options)
   at Microsoft.Extensions.Options.OptionsFactory`1.Create(String name)
   at Microsoft.Extensions.Options.OptionsMonitor`1.<>c.<Get>b__10_0(String name, IOptionsFactory`1 factory)
   at Microsoft.Extensions.Options.OptionsCache`1.<>c__DisplayClass3_1`1.<GetOrAdd>b__2()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at Microsoft.Extensions.Options.OptionsCache`1.GetOrAdd[TArg](String name, Func`3 createOptions, TArg factoryArgument)
   at Microsoft.Extensions.Options.OptionsMonitor`1.Get(String name)
   at Microsoft.AspNetCore.Authentication.AuthenticationHandler`1.InitializeAsync(AuthenticationScheme scheme, HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationHandlerProvider.GetHandlerAsync(HttpContext context, String authenticationScheme)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
```

Reviewer: Running on 4x now returns the following error:

> System.InvalidOperationException: IDX20803: Unable to obtain configuration from: 'http://localhost:8080/uaa/.well-known/openid-configuration'. Will retry at '1/1/0001 12:00:00 AM +00:00'. Exception: 'System.IO.IOException: IDX20804: Unable to retrieve document from: 'http://localhost:8080/uaa/.well-known/openid-configuration'.

I'm using the UAA Steeltoe Docker container.

Author: Required for local UAA:
```json
"MetadataAddress": "http://localhost:8080/.well-known/openid-configuration",
"RequireHttpsMetadata": false,
```

builder.Services.AddAuthentication((options) =>
    {
        options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
-        options.DefaultChallengeScheme = CloudFoundryDefaults.AuthenticationScheme;
+        options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
    })
    .AddCookie(options => options.AccessDeniedPath = new PathString("/Home/AccessDenied"))
-    .AddCloudFoundryOAuth(builder.Configuration);
-    .AddCloudFoundryOpenIdConnect(builder.Configuration);
+    .AddOpenIdConnect().ConfigureOpenIdConnectForCloudFoundry();
builder.Services.AddAuthorizationBuilder()
    .AddPolicy("some-group", policy => policy.RequireClaim("scope", "some-group"));

var app = builder.Build();

-app.UseForwardedHeaders(new ForwardedHeadersOptions
-{
-    ForwardedHeaders = ForwardedHeaders.XForwardedProto
-});

app.UseRouting();

app.UseAuthentication();
app.UseAuthorization();
```

appsettings.json:

```diff
-  "$schema": "https://steeltoe.io/schema/v3/schema.json",
-  "Security": {
-    "Oauth2": {
-      "Client": {
-        "OAuthServiceUrl": "http://localhost:8080/uaa",
-        "ClientId": "steeltoesamplesclient",

> -        "ClientSecret": "client_secret",

Reviewer: I'm not sure if this should have been addressed by now, but there are still some left.

var app = builder.Build();

-app.UseForwardedHeaders(new ForwardedHeadersOptions
-{
-    ForwardedHeaders = ForwardedHeaders.XForwardedProto
-});

app.UseRouting();

app.UseAuthentication();
app.UseAuthorization();
```

> [!NOTE]
> The code above should also be used for applications that previously used `.AddCloudFoundryOAuth(builder.Configuration);`

### JWT Bearer

Project file:

```diff
<Project>
  <ItemGroup>
-    <PackageReference Include="Steeltoe.Extensions.Configuration.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Configuration.CloudFoundry" Version="4.0.0" />
-    <PackageReference Include="Steeltoe.Security.Authentication.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Security.Authentication.JwtBearer" Version="4.0.0" />
  </ItemGroup>
</Project>
```

appsettings.json:

```diff
-  "$schema": "https://steeltoe.io/schema/v3/schema.json",
-  "Security": {
-    "Oauth2": {
-      "Client": {
-        "OAuthServiceUrl": "http://localhost:8080/uaa",
-        "ClientId": "steeltoesamplesserver",
-        "ClientSecret": "server_secret",

> -      }

Author: Until this pr merges, something like the following will be required for Jwt to work:

```json
        "tokenvalidationparameters": {
          "validAudience": "sampleapi",
           // or
          "validaudiences": [ "sampleapi", "steeltoesamplesclient" ]
        }
```

The CredHub client has been removed from Steeltoe in v4.
Use [CredHub Service Broker](https://techdocs.broadcom.com/us/en/vmware-tanzu/platform-services/credhub-service-broker/services/credhub-sb/index.html) instead.

### OAuth and OpenID Connect

OAuth support has been removed from Steeltoe in v4. Use OpenID Connect instead.

Project file:

```diff
<Project>
  <ItemGroup>
-    <PackageReference Include="Steeltoe.Extensions.Configuration.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Configuration.CloudFoundry" Version="4.0.0" />
-    <PackageReference Include="Steeltoe.Security.Authentication.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Security.Authentication.OpenIdConnect" Version="4.0.0" />
  </ItemGroup>
</Project>
```

appsettings.json:

```diff
-  "$schema": "https://steeltoe.io/schema/v3/schema.json",
-  "Security": {
-    "Oauth2": {

> -      "Client": {

Author: The schema is super wrong. It's based on `OAuthConnectorOptions` which is bound from a very different path (`VCAP_SERVICES` or the local equivalent).

For this path, the Options class is actually [`CloudFoundryOAuthOptions`](https://github.com/SteeltoeOSS/Steeltoe/blob/3.x/src/Security/src/Authentication.CloudFoundryCore/CloudFoundryOAuthOptions.cs), [`CloudFoundryOpenIdConnectOptions`](https://github.com/SteeltoeOSS/Steeltoe/blob/3.x/src/Security/src/Authentication.CloudFoundryCore/CloudFoundryOpenIdConnectOptions.cs) or [`CloudFoundryJwtBearerOptions`](https://github.com/SteeltoeOSS/Steeltoe/blob/3.x/src/Security/src/Authentication.CloudFoundryCore/CloudFoundryJwtBearerOptions.cs), with all 3 cases getting the majority of their properties from underlying MSFT classes. I'm going to separate OAuth from OIDC and describe moving from OAuth to OIDC as a precursor to upgrading to v4.

Additionally, for OAuth and JWT we need to use the key `AuthDomain` locally because of [this line](https://github.com/SteeltoeOSS/Steeltoe/blob/3.x/src/Security/src/Authentication.CloudFoundryCore/AuthenticationBuilderExtensions.cs#L87) (which jumps [to this line](https://github.com/SteeltoeOSS/Steeltoe/blob/3.x/src/Security/src/Authentication.CloudFoundryCore/AuthenticationBuilderExtensions.cs#L293)). 

Short story _very_ long, most of the auth pieces _really_ weren't designed or documented for running locally

> -      "Client": {

Reviewer: Line 293 doesn't exist anymore :)

Now that this section describes how to upgrade from OIDC@v3 to OIDC@v4, why is the old configuration still using the `Security:Oauth2:Client` path?

-using Microsoft.AspNetCore.HttpOverrides;
-using Steeltoe.Extensions.Configuration.CloudFoundry;
+using Steeltoe.Configuration.CloudFoundry;
-using Steeltoe.Security.Authentication.CloudFoundry;
+using Steeltoe.Security.Authentication.JwtBearer;

var builder = WebApplication.CreateBuilder(args);

builder.AddCloudFoundryConfiguration();

builder.Services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
-    .AddCloudFoundryJwtBearer(builder.Configuration);
+    .AddJwtBearer().ConfigureJwtBearerForCloudFoundry();
builder.Services.AddAuthorizationBuilder()
    .AddPolicy("some-group", policy => policy.RequireClaim("scope", "some-group"));

var app = builder.Build();

-app.UseForwardedHeaders(new ForwardedHeadersOptions
-{
-    ForwardedHeaders = ForwardedHeaders.XForwardedProto
-});

app.UseAuthentication();
app.UseAuthorization();
```

### Certificates / MutualTLS

Project file:

```diff
<Project>
  <ItemGroup>
-    <PackageReference Include="Steeltoe.Security.Authentication.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Security.Authorization.Certificate" Version="4.0.0" />
  </ItemGroup>
</Project>
```

> Program.cs (server-side):

Reviewer: Can you follow the same approach as I used for discovery? Define an inline minimal-API endpoint and use a typed HttpClient to call it.

-app.UseCloudFoundryCertificateAuth();
+app.UseCertificateAuthorization();
```

Secured endpoints:

```diff
-using Steeltoe.Security.Authentication.CloudFoundry;
+using Steeltoe.Security.Authorization.Certificate;

-        [Authorize(CloudFoundryDefaults.SameOrganizationAuthorizationPolicy)]
+        [Authorize(CertificateAuthorizationPolicies.SameOrg)]
        [HttpGet]
        public string SameOrg()
        {
            return "Certificate is valid and both client and server are in the same org";
        }

-        [Authorize(CloudFoundryDefaults.SameSpaceAuthorizationPolicy)]
+        [Authorize(CertificateAuthorizationPolicies.SameSpace)]
        [HttpGet]
        public string SameSpace()
        {
            return "Certificate is valid and both client and server are in the same space";
        }
```

Program.cs (client-side):

```diff
-using System.Security.Cryptography.X509Certificates;
-using Microsoft.Extensions.Options;
-using Steeltoe.Common.Options;
+using Steeltoe.Common.Certificates;
-using Steeltoe.Security.Authentication.CloudFoundry;
+using Steeltoe.Security.Authorization.Certificate;

var builder = WebApplication.CreateBuilder(args);

> -builder.Configuration.AddCloudFoundryContainerIdentity();

Author: `-builder.Services.AddCloudFoundryContainerIdentity()` was missing

The CredHub client has been removed from Steeltoe in v4.
Use [CredHub Service Broker](https://techdocs.broadcom.com/us/en/vmware-tanzu/platform-services/credhub-service-broker/services/credhub-sb/index.html) instead.

### OAuth

OAuth support has been removed from Steeltoe in v4. Use OpenID Connect instead.

> Consider migrating to OpenID Connect on Steeltoe v3 before moving to v4.

Reviewer: ```suggestion
Before migrating to Steeltoe v4, apply the following changes to migrate from OAuth to OpenID Connect using Steeltoe v3.
```

Assuming these are the steps needed to migrate OAuth in v3 to OIDC in v3. I'm not sure if that's the intent.

It's not something to _consider_ before upgrade, as there's no migration possible to v4 if the user wants to keep using OAuth.

appsettings.json:

```diff
{
  "$schema": "https://steeltoe.io/schema/v3/schema.json",
  "Security": {
    "Oauth2": {
      "Client": {
-        "AuthDomain": "http://localhost:8080",
+        "Authority": "http://localhost:8080/uaa",
        "CallbackPath": "/signin-oidc",
        "ClientId": "steeltoesamplesclient",
        "ClientSecret": "client_secret",
+        "MetadataAddress": "http://localhost:8080/.well-known/openid-configuration",
+        "RequireHttpsMetadata": false,

> +        "AdditionalScopes": "sampleapi.read"

Reviewer: Can these be moved right after "Authority", to avoid the trailing comma issue?

Reviewer: This applies to multiple places. Please use the same ordering throughout the config files where possible, to make it easier to spot differences.

app.UseForwardedHeaders(new ForwardedHeadersOptions
{
    ForwardedHeaders = ForwardedHeaders.XForwardedProto
});

app.UseAuthentication();
app.UseAuthorization();

```

### OpenID Connect

Project file:

```diff
<Project>
  <ItemGroup>
-    <PackageReference Include="Steeltoe.Extensions.Configuration.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Configuration.CloudFoundry" Version="4.0.0" />
-    <PackageReference Include="Steeltoe.Security.Authentication.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Security.Authentication.OpenIdConnect" Version="4.0.0" />
  </ItemGroup>
</Project>
```

appsettings.json:

```diff
{
-  "$schema": "https://steeltoe.io/schema/v3/schema.json",
+  "$schema": "https://steeltoe.io/schema/v4/schema.json",
-  "Security": {
-    "Oauth2": {
-      "Client": {
-        "Authority": "http://localhost:8080/uaa",
-        "CallbackPath": "/signin-oidc",
-        "ClientId": "steeltoesamplesclient",
-        "ClientSecret": "client_secret",
-        "MetadataAddress": "http://localhost:8080/.well-known/openid-configuration",
-        "AdditionalScopes": "sampleapi.read",
-        "SaveTokens": true,

> -        "RequireHttpsMetadata": false

Reviewer: Please move `"RequireHttpsMetadata"` to right after `"MetadataAddress"`, so it's easier to spot how it differs from the new settings. For the same reason, move `"AdditionalScopes"` to the end.

-    .AddCloudFoundryOpenIdConnect(builder.Configuration);
+    .AddOpenIdConnect().ConfigureOpenIdConnectForCloudFoundry();
builder.Services.AddAuthorizationBuilder()
    .AddPolicy("read", policy => policy.RequireClaim("scope", "sampleapi.read"))
    .AddPolicy("write", policy => policy.RequireClaim("scope", "sampleapi.write"));

var app = builder.Build();

-app.UseForwardedHeaders(new ForwardedHeadersOptions
-{
-    ForwardedHeaders = ForwardedHeaders.XForwardedProto
-});

app.UseAuthentication();
app.UseAuthorization();
```

### JWT Bearer

Project file:

```diff
<Project>
  <ItemGroup>
-    <PackageReference Include="Steeltoe.Extensions.Configuration.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Configuration.CloudFoundry" Version="4.0.0" />
-    <PackageReference Include="Steeltoe.Security.Authentication.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Security.Authentication.JwtBearer" Version="4.0.0" />
  </ItemGroup>
</Project>
```

appsettings.json:

```diff
{
-  "$schema": "https://steeltoe.io/schema/v3/schema.json",
+  "$schema": "https://steeltoe.io/schema/v4/schema.json",
-  "Security": {
-    "Oauth2": {

> -      "Client": {

Reviewer: Should OAuth2 be adjusted for JWT here?

-});

app.UseAuthentication();
app.UseAuthorization();

app.MapGet("/jwt", async httpContext =>
    {
        httpContext.Response.StatusCode = 200;
        httpContext.Response.ContentType = "text/plain";
        await httpContext.Response.WriteAsync("JWT is valid and contains the required claim");
    }).RequireAuthorization("sampleapi.read");
app.Run();
```

### Client Certificates (Mutual TLS)

Project file:

```diff
<Project>
  <ItemGroup>
-    <PackageReference Include="Steeltoe.Security.Authentication.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Security.Authorization.Certificate" Version="4.0.0" />
  </ItemGroup>
</Project>
```

Program.cs (server-side):

```diff
+using Steeltoe.Common.Certificates;
-using Steeltoe.Security.Authentication.CloudFoundry;
+using Steeltoe.Security.Authorization.Certificate;

const string orgId = "a8fef16f-94c0-49e3-aa0b-ced7c3da6229";
const string spaceId = "122b942a-d7b9-4839-b26e-836654b9785f";

var builder = WebApplication.CreateBuilder(args);

-builder.Configuration.AddCloudFoundryContainerIdentity(orgId, spaceId);
+builder.Configuration.AddAppInstanceIdentityCertificate(new Guid(orgId), new Guid(spaceId));

> -builder.Services.AddCloudFoundryCertificateAuth(options => options.CertificateHeader = "X-Client-Cert");

Reviewer: I recall that we were debugging and found the key needed to be changed to `X-Forward-Client-Cert` (or something like that). Please dismiss if this is intentional and I'm mixing things up.

    {
        httpContext.Response.StatusCode = 200;
        httpContext.Response.ContentType = "text/plain";
        await httpContext.Response.WriteAsync("Client and server identity certificates have matching Org values");
    })
-    .RequireAuthorization(CloudFoundryDefaults.SameOrganizationAuthorizationPolicy);
+    .RequireAuthorization(CertificateAuthorizationPolicies.SameOrg);

app.MapGet("/sameSpace", async httpContext =>
    {
        httpContext.Response.StatusCode = 200;
        httpContext.Response.ContentType = "text/plain";
        await httpContext.Response.WriteAsync("Client and server identity certificates have matching Space values");
    })
-    .RequireAuthorization(CloudFoundryDefaults.SameSpaceAuthorizationPolicy);
+    .RequireAuthorization(CertificateAuthorizationPolicies.SameSpace);
```

Program.cs (client-side):

```diff
-using System.Security.Cryptography.X509Certificates;
-using Microsoft.Extensions.Options;
-using Steeltoe.Common.Options;
+using Steeltoe.Common.Certificates;
-using Steeltoe.Security.Authentication.CloudFoundry;
+using Steeltoe.Security.Authorization.Certificate;

const string orgId = "a8fef16f-94c0-49e3-aa0b-ced7c3da6229";

> const string spaceId = "122b942a-d7b9-4839-b26e-836654b9785f";

Reviewer: A note at the bottom of the section would be helpful to explain that this is only for local development to ensure the same values are used by server and client, and that these GUIDs must be omitted when running on Cloud Foundry.

Consider adding `// see note below` in the code.

var builder = WebApplication.CreateBuilder(args);

-builder.Configuration.AddCloudFoundryContainerIdentity(orgId, spaceId);
+builder.Configuration.AddAppInstanceIdentityCertificate(new Guid(orgId), new Guid(spaceId));
-builder.Services.AddCloudFoundryContainerIdentity();
builder.Services
-    .AddHttpClient<PingClient>((services, client) =>
-{
-    client.BaseAddress = new Uri("http://example-service/")
-    var options = services.GetRequiredService<IOptions<CertificateOptions>>();
-    var b64 = Convert.ToBase64String(options.Value.Certificate.Export(X509ContentType.Cert));
-    client.DefaultRequestHeaders.Add("X-Client-Cert", b64);
-});

> +    .AddHttpClient<PingClient>(httpClient => httpClient.BaseAddress = new Uri("http://example-service/"))

Reviewer: `http://example-service/` won't work here, as we have no discovery. We could put a random port number here, but a more elegant approach would be to add a `launchsettings.json` that defines the host/port, so that readers can relate to where this is coming from.

+        .AddAppInstanceIdentityCertificate();

var app = builder.Build();

> var pingClient = app.Services.GetRequiredService<PingClient>();

Reviewer: What's the point of the client calling itself? I would expect a call to the `/sameOrg` and `/sameSpace` endpoints on the server here.

var app = builder.Build();

app.UseForwardedHeaders(new ForwardedHeadersOptions
{
    ForwardedHeaders = ForwardedHeaders.XForwardedProto
});

app.UseAuthentication();
app.UseAuthorization();

```

### OpenID Connect

Project file:

```diff
<Project>
  <ItemGroup>
-    <PackageReference Include="Steeltoe.Extensions.Configuration.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Configuration.CloudFoundry" Version="4.0.0" />
-    <PackageReference Include="Steeltoe.Security.Authentication.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Security.Authentication.OpenIdConnect" Version="4.0.0" />
  </ItemGroup>
</Project>
```

appsettings.json:

```diff
{
-  "$schema": "https://steeltoe.io/schema/v3/schema.json",
+  "$schema": "https://steeltoe.io/schema/v4/schema.json",
-  "Security": {
-    "Oauth2": {
-      "Client": {
-        "Authority": "http://localhost:8080/uaa",
-        "CallbackPath": "/signin-oidc",
-        "ClientId": "steeltoesamplesclient",
-        "ClientSecret": "client_secret",
-        "MetadataAddress": "http://localhost:8080/.well-known/openid-configuration",
-        "RequireHttpsMetadata": false,

> -        "SaveTokens": true,

Author: `SaveTokens` was there for passing the token to a backing service that's JWT-secured, but as that ultimately isn't mentioned in this guide I'll just remove it entirely.

> -        "SaveTokens": true,

Author: This didn't sit right with me, so I went back and looked at why setting `SaveTokens` wasn't needed when using OAuth and found [this line in Steeltoe](https://github.com/SteeltoeOSS/Steeltoe/blob/3.x/src/Security/src/Authentication.CloudFoundryCore/CloudFoundryOAuthOptions.cs#L42) that sets it to true by default. So that at least explains what's happening, not sure if it's worth adding here?

+        "RequireHttpsMetadata": false,
+        "ValidAudiences": [ "sampleapi" ]
+      }
+    }
+  }
}
```

> [!NOTE]
> This is not a complete listing of appsettings. As of version 4, Steeltoe configures Microsoft's option class rather than maintaining separate options.
> Refer to [the JwtBearerOptions class documentation](https://learn.microsoft.com/dotnet/api/microsoft.aspnetcore.authentication.jwtbearer.jwtbeareroptions) for the new options.

Program.cs:

```diff
-using Microsoft.AspNetCore.HttpOverrides;
-using Steeltoe.Extensions.Configuration.CloudFoundry;
+using Steeltoe.Configuration.CloudFoundry;
+using Steeltoe.Configuration.CloudFoundry.ServiceBindings;
-using Steeltoe.Security.Authentication.CloudFoundry;
+using Steeltoe.Security.Authentication.JwtBearer;

var builder = WebApplication.CreateBuilder(args);

builder.AddCloudFoundryConfiguration();
+builder.Configuration.AddCloudFoundryServiceBindings();

builder.Services.AddAuthentication()
-    .AddCloudFoundryJwtBearer(builder.Configuration);
+    .AddJwtBearer().ConfigureJwtBearerForCloudFoundry();
builder.Services.AddAuthorizationBuilder()

>     .AddPolicy("sampleapi.read", policy => policy.RequireClaim("scope", "sampleapi.read"));

Author: `sampleapi.write` was there to define a policy where the user isn't authorized, I added it for a quick demo where you were experiencing failures walking through this... for this guide we ultimately don't need to keep it

var app = builder.Build();

app.UseForwardedHeaders(new ForwardedHeadersOptions
{
    ForwardedHeaders = ForwardedHeaders.XForwardedProto
});

app.UseAuthentication();
app.UseAuthorization();

app.MapGet("/test-auth", async httpContext =>
    {
        httpContext.Response.StatusCode = 200;
        httpContext.Response.ContentType = "text/plain";
        await httpContext.Response.WriteAsync("You are logged in and carry the required claim.");
    }).RequireAuthorization("read");
```

### OpenID Connect

Project file:

```diff
<Project>
  <ItemGroup>
-    <PackageReference Include="Steeltoe.Extensions.Configuration.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Configuration.CloudFoundry" Version="4.0.0" />
-    <PackageReference Include="Steeltoe.Security.Authentication.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Security.Authentication.OpenIdConnect" Version="4.0.0" />
  </ItemGroup>
</Project>
```

appsettings.json:

```diff
{
-  "$schema": "https://steeltoe.io/schema/v3/schema.json",
+  "$schema": "https://steeltoe.io/schema/v4/schema.json",
-  "Security": {
-    "Oauth2": {
-      "Client": {

> -        "Authority": "http://localhost:8080",

Author: Yes, `/uaa` isn't needed with the new UAA image

+builder.Configuration.AddCloudFoundryServiceBindings();
builder.Services.AddAuthentication(options =>
    {
        options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
-        options.DefaultChallengeScheme = CloudFoundryDefaults.AuthenticationScheme;
+        options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
    })
    .AddCookie(options => options.AccessDeniedPath = new PathString("/Home/AccessDenied"))
-    .AddCloudFoundryOpenIdConnect(builder.Configuration);
+    .AddOpenIdConnect().ConfigureOpenIdConnectForCloudFoundry();
builder.Services.AddAuthorizationBuilder()
    .AddPolicy("read", policy => policy.RequireClaim("scope", "sampleapi.read"));

var app = builder.Build();

-app.UseForwardedHeaders(new ForwardedHeadersOptions
-{
-    ForwardedHeaders = ForwardedHeaders.XForwardedProto
-});

app.UseAuthentication();
app.UseAuthorization();

app.MapGet("/test-auth", async httpContext =>
    {
        httpContext.Response.StatusCode = 200;
        httpContext.Response.ContentType = "text/plain";
        await httpContext.Response.WriteAsync("You are logged in and carry the required claim.");

>     }).RequireAuthorization("read");

Reviewer: ```suggestion
    }).RequireAuthorization("read");

app.Run();    
```

```

### JWT Bearer

Project file:

```diff
<Project>
  <ItemGroup>
-    <PackageReference Include="Steeltoe.Extensions.Configuration.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Configuration.CloudFoundry" Version="4.0.0" />
-    <PackageReference Include="Steeltoe.Security.Authentication.CloudFoundryCore" Version="3.*" />
+    <PackageReference Include="Steeltoe.Security.Authentication.JwtBearer" Version="4.0.0" />
  </ItemGroup>
</Project>
```

appsettings.json:

```diff
{
-  "$schema": "https://steeltoe.io/schema/v3/schema.json",
+  "$schema": "https://steeltoe.io/schema/v4/schema.json",
-  "Security": {
-    "Oauth2": {

> -      "Client": {

Author: Yes. Previously it was needed to override the calculated path of `/uaa/token_keys`

-using Steeltoe.Extensions.Configuration.CloudFoundry;
+using Steeltoe.Configuration.CloudFoundry;
+using Steeltoe.Configuration.CloudFoundry.ServiceBindings;
-using Steeltoe.Security.Authentication.CloudFoundry;
+using Steeltoe.Security.Authentication.JwtBearer;

var builder = WebApplication.CreateBuilder(args);

builder.AddCloudFoundryConfiguration();
+builder.Configuration.AddCloudFoundryServiceBindings();
builder.Services.AddAuthentication()
-    .AddCloudFoundryJwtBearer(builder.Configuration);
+    .AddJwtBearer().ConfigureJwtBearerForCloudFoundry();
builder.Services.AddAuthorizationBuilder()
    .AddPolicy("sampleapi.read", policy => policy.RequireClaim("scope", "sampleapi.read"));

var app = builder.Build();

-app.UseForwardedHeaders(new ForwardedHeadersOptions
-{
-    ForwardedHeaders = ForwardedHeaders.XForwardedProto
-});

app.UseAuthentication();
app.UseAuthorization();

app.MapGet("/test-jwt", async httpContext =>
    {
        httpContext.Response.StatusCode = 200;
        httpContext.Response.ContentType = "text/plain";
        await httpContext.Response.WriteAsync("JWT is valid and contains the required claim.");
    }).RequireAuthorization("sampleapi.read");

> app.Run();

Reviewer: ```suggestion

app.Run();
```

+    <PackageReference Include="Steeltoe.Security.Authorization.Certificate" Version="4.0.0" />
  </ItemGroup>
</Project>
```

launchsettings.json (server-side):

```diff
{
  "profiles": {
    "http": {
      "commandName": "Project",
      "applicationUrl": "https://localhost:7107"
    }
  }
}
```

Program.cs (server-side):

```diff
+using Steeltoe.Common.Certificates;
-using Steeltoe.Security.Authentication.CloudFoundry;
+using Steeltoe.Security.Authorization.Certificate;

const string orgId = "a8fef16f-94c0-49e3-aa0b-ced7c3da6229";
const string spaceId = "122b942a-d7b9-4839-b26e-836654b9785f";

var builder = WebApplication.CreateBuilder(args);
-builder.Configuration.AddCloudFoundryContainerIdentity(orgId, spaceId);
+builder.Configuration.AddAppInstanceIdentityCertificate(new Guid(orgId), new Guid(spaceId));
-builder.Services.AddCloudFoundryCertificateAuth(options => options.CertificateHeader = "X-Forwarded-Client-Cert");
+builder.Services.AddAuthentication().AddCertificate();
+builder.Services.AddAuthorizationBuilder().AddOrgAndSpacePolicies("X-Forwarded-Client-Cert");

var app = builder.Build();

-app.UseCloudFoundryCertificateAuth();
+app.UseCertificateAuthorization();

> app.MapGet("/test-same-org", async httpContext =>

Reviewer: ```suggestion

app.MapGet("/test-same-org", async httpContext =>
```

    {
        httpContext.Response.StatusCode = 200;
        httpContext.Response.ContentType = "text/plain";
        await httpContext.Response.WriteAsync("Client and server identity certificates have matching Org values.");
    })
-    .RequireAuthorization(CloudFoundryDefaults.SameOrganizationAuthorizationPolicy);
+    .RequireAuthorization(CertificateAuthorizationPolicies.SameOrg);
app.MapGet("/test-same-space", async httpContext =>
    {
        httpContext.Response.StatusCode = 200;
        httpContext.Response.ContentType = "text/plain";
        await httpContext.Response.WriteAsync("Client and server identity certificates have matching Space values.");
    })
-    .RequireAuthorization(CloudFoundryDefaults.SameSpaceAuthorizationPolicy);

> +    .RequireAuthorization(CertificateAuthorizationPolicies.SameSpace);

Reviewer: ```suggestion
+    .RequireAuthorization(CertificateAuthorizationPolicies.SameSpace);

app.Run();
```
